### DISCUSSION TITLE
A question about the template class MooseVariableInterface

### DISCUSSION THREAD

**Discussion by TJT-post95** [idaholab/moose#22278] [Q&A General] [answered]
`MooseVariableInterface.h` is included and inherited in the `Kernel` class. `MooseVariableInterface` is a template class, and the code is written in two files. However, from my understanding, this approach usually does not work properly. Because for the template class, maybe all the code should be written in the `MooseVariableInterface.h`, or `MooseVariableInterface.C` is supposed to be included. So, I'm curious if MOOSE does any additional setup?
Thank you!

**Comment by GiudGiud** [comment, marked as answer]
Hello

We explicitly instantiate every template in the source file for that template. 
It's not the most portable option, as someone compiling a new app with a new instantiation needed would need to modify MOOSE, but that can be an advantage too. It works for many of our templates because we know what we support and this is helpful in disallowing what we do not support. 

For example for MooseVariableInterface,
```
template class MooseVariableInterface<Real>;
template class MooseVariableInterface<RealVectorValue>;
template class MooseVariableInterface<RealEigenVector>;
```
we know we would not support `MooseVariableInterface<Matrix>;` without a lot of work elsewhere in the framework to handle the matrices, even though that particular template would compile just fine. 

Guillaume

**Comment by TJT-post95** [comment]
Got! Thank you!
